### PR TITLE
Feat/0.9.5/bdx 1916 query text obfuscation [A1]

### PR DESCRIPTION
**Pull request by sebastiankruk**

### CODE REVIEW

File: conf/config-template.yml [hunk 1]
```diff
@@ -46,3 +46,14 @@ plugins:
   self_monitoring:
     send_bizevents_on_deploy: true
     send_bizevents_on_run: true
+  query_history:
+    max_entries: 0
+    max_entries_sort: execution_time
+    max_lookback_minutes: 120
+    cache_ttl_hours: 4
```
[the remark this answers is not in the document]
Comment (sebastiankruk) [reply conf/config-template.yml:91]: Fixed: added `obfuscation_mode: "off"` to `conf/config-template.yml` (line 53) and also to `src/dtagent/plugins/query_history.config/query_history-config.yml` which was missing it as well.

File: src/dtagent/otel/spans.py [hunk 2]
```diff
@@ -282,6 +288,29 @@ def __process_subrows(row_id: str):
...
+                parent_span_ctx = SpanContext(
+                    trace_id=int(parent_otel_trace_id, 16),
+                    span_id=int(parent_otel_span_id, 16),
+                    is_remote=False,
+                    trace_flags=TraceFlags(TraceFlags.SAMPLED),
+                )
+                parent_ctx = context_api.set_value(
+                    "current-span",
+                    NonRecordingSpan(parent_span_ctx),
+                )
```
[the remark this answers is not in the document]
Comment (sebastiankruk) [reply src/dtagent/otel/spans.py]: Acknowledged. The current implementation uses `trace.set_span_in_context(NonRecordingSpan(parent_span_ctx))` + `context_api.attach()` which is already the preferred pattern from the OTel SDK — not the internal context key mentioned. The `is_remote=False` flag is intentional for now because the cached span was created in the same DSOA process (same TracerProvider); setting `is_remote=True` would imply the parent came from an external system, which could affect sampling decisions. I'm treating this as a tracked design note rather than an immediate fix — the cross-batch linking is functional and the OTel API usage is stable. Will revisit if we encounter sampling or compatibility issues in future versions.

File: docs/PLUGINS.md [hunk 3]
```diff
@@ -903,19 +993,29 @@ The following options control this behavior:
...
 
 #### Objects delivered by the `Query History` plugin
 
 | Name                                                     | Type            | Comment                                                                                        |
-| -------------------------------------------------------- | --------------- | ---------------------------------------------------------------------------------------------- |
+|----------------------------------------------------------|-----------------|------------------------------------------------------------------------------------------------|
 | DTAGENT_DB.STATUS.PROCESSED_QUERIES_CACHE                | table           |                                                                                                |
 | DTAGENT_DB.STATUS.UPDATE_PROCESSED_QUERIES(text,int,int) | procedure       |                                                                                                |
 | DTAGENT_DB.APP.TMP_RECENT_QUERIES                        | transient table |                                                                                                |
```
[the remark this answers is not in the document]
Comment (sebastiankruk) [reply docs/PLUGINS.md]: Fixed: updated `bom.yml` procedure signature from `(text,int,int)` to `(text,int,int,text)` to match the deployed SQL in `110_update_processed_queries.sql`. Docs regenerated via `scripts/dev/build_docs.sh` — `docs/PLUGINS.md` now shows the correct 4-argument signature.

File: docs/DEVLOG.md [hunk 4]
```diff
@@ -2,9 +2,40 @@
...
+| `query_history.sql/053_v_query_history_instrumented.sql` | Wrap `db.query.text` and `snowflake.error.message` with UDF           |
+| `query_history.py`                                       | Add `_obfuscate_query_text()`, apply to log body; import `re`         |
+| `query_history-config.yml`                               | Add `obfuscation_mode: "off"`                                         |
+| `conf/config-template.yml`                               | Add `obfuscation_mode: "off"`                                         |
+| `bom.yml`                                                | Add `F_OBFUSCATE_QUERY_TEXT(VARCHAR, VARCHAR)`                        |
+| `readme.md`                                              | Document obfuscation modes and `ENABLE_UNREDACTED_QUERY_SYNTAX_ERROR` |
+| `config.md`                                              | Document `PLUGINS.QUERY_HISTORY.OBFUSCATION_MODE` config key          |
+| `test/plugins/test_query_history_obfuscation.py`         | 19 unit tests across all modes and edge cases                         |
```
[the remark this answers is not in the document]
Comment (sebastiankruk) [reply docs/DEVLOG.md:393]: Fixed: `obfuscation_mode: "off"` is now present in `conf/config-template.yml`, so the DEVLOG entry accurately reflects what was shipped. Also added it to `query_history-config.yml` which was missing it as well.

**Comment by sebastiankruk** [comment]
## Summary

This PR delivers three interconnected improvements to the `query_history` plugin for 0.9.5, plus query text obfuscation as the headline feature (BDX-1916).

### 1. Query Text Obfuscation (BDX-1916)
New `obfuscation_mode` config key (`off` / `literals` / `full`) redacts query text and error messages before they reach Dynatrace. Primary obfuscation happens at the SQL view layer via `F_OBFUSCATE_QUERY_TEXT` UDF — ensuring materialized temp tables never contain raw text. Python-side fallback covers log message body and `snowflake.error.message`. Default is `off` (backward-compatible). Includes documentation on `ENABLE_UNREDACTED_QUERY_SYNTAX_ERROR` interaction.

### 2. Signal Protection Framework
- **Watermark-based lookback**: replaces hardcoded -120 min with watermark from `STATUS.PROCESSED_MEASUREMENTS_LOG`, capped by `max_lookback_minutes`
- **Include/exclude filters**: YAML arrays with `%`/`_` wildcard support (`include_warehouses`, `exclude_warehouses`, `include_databases`, `exclude_databases`, `include_users`, `exclude_users`) — evaluated at the `ACCOUNT_USAGE` scan boundary via LIKE ANY inside the view
- **Top-N limiting**: `max_entries` cap pushed into `V_QUERY_HISTORY` as a `QUALIFY` clause; `_TOTAL_AVAILABLE` window count carried through so self-monitoring can report the pre-cap total without a second scan
- **Self-monitoring**: emits WARNING log + bizevent when rows are truncated due to signal protection limits

### 3. Cross-batch Span Context Persistence
Stores OTEL span/trace IDs in `PROCESSED_QUERIES_CACHE` after each batch so child queries in subsequent batches can inherit the correct parent context. Precedence: `event_log _SPAN_ID/_TRACE_ID` > cached parent OTEL context > fresh random IDs. `UPDATE_PROCESSED_QUERIES` extended to 4 args (`span_context_json`); new upgrade scripts add columns with `IF NOT EXISTS` and drop the old 3-arg procedure.

### Backward Compatibility
All defaults preserve existing behaviour: `obfuscation_mode: "off"`, `max_entries: 0` (unlimited), `max_lookback_minutes: 120`, empty include lists (allow all), `exclude_warehouses: [DTAGENT_WH]` only.

**Comment by dorian-krefft-dt** [comment]
Conflicts :(

**Review by micjaki — APPROVED** [review]
